### PR TITLE
Added support for multiple SQL Database CreateModes (including geo-replication) and other extension methods

### PR DESCRIPTION
**Pull request by LukeHarjulin**
**Progression with SQL Database**

Database-based geo-location works through the extension method SetAsSecondary. It sets the database to be a geo-replica of a primary db (of which the id is passed into the SetAsSecondary method)

Added more CreateMode options:
- to Restore a deleted db - use SetAsRestore()
- to "restore a geo-replication of backup" - use SetAsRecovery() (see [Link](https://www.pulumi.com/docs/reference/pkg/azure-native/sql/database/#createmode_csharp))
- to copy a db - use SetAsCopy()
- to restore a database from long term retention - use SetAsLongTermRetentionRestore()
- to restore from a point in time backup - use SetAsPointInTimeRestore()
- to set the db as a geo-replica backup (secondary) - use SetAsSecondary()

Added more ext methods:
- SecondaryType() -> not entirely sure what this is, but the default value is 'geo' (geo-replication)
- EnableZoneRedundancy() -> by default Zone Redundancy is disabled, so I made an option to enable it
- MaxDatabaseSizeGB() -> set the database size in GB with an integer (method converts to bytes and inserts into resource argument 'MaxDatabaseSizeBytes') 
- MinCapacity() -> I've not figured out what this does yet. 
- StorageAccountType -> Takes StorageAccountType struct as an argument (valid values: GRS, LRS, ZRS)
- DatabaseCollation() -> Set the collation of the database
- SampleData() -> to insert sample data into the database

Added corresponding unit test methods to test _most_ of the new extension methods (some can't be unit tested).

Most new extension methods have been tested through deployment using 'pulumi up'.

### CODE REVIEW

File: src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs [hunk 1]
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
...
+        }
+
+        /// <summary>
+        /// Sets the database to be a copy of an existing database
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="databaseId">Resource ID of the database to copy</param>
         /// <returns></returns>
-        public static SqlDatabaseBuilder RestoreFrom(this SqlDatabaseBuilder builder, Input<string> databaseId)
+        public static SqlDatabaseBuilder SetAsCopy(this SqlDatabaseBuilder builder, Input<string> databaseId)
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:133]: Should we rename this to "CreateAsCopy"? I feel the current name is a bit confusing.

File: src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs [hunk 2]
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
...
         }
 
         /// <summary>
-        /// Enables the read scale
+        /// Sets the database to be a restoration of a geo-replicated backup database.
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="databaseId">Recoverable Database Resource ID of the database to restore</param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder SetAsRecovery(this SqlDatabaseBuilder builder, Input<string> databaseId)
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:146]: Should we rename this to "CreateAsRecoveryOf"? I feel the current name is a bit confusing.

File: src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs [hunk 3]
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
...
+        }
+
+        /// <summary>
+        /// Sets the database to be a restoration of a point in time of an existing database, using the existing database's resource ID and the point in time (ISO8601 format) to restore from.
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="databaseId">Resource ID of the database to restore from</param>
+        /// <param name="restorePointInTime">The point in time (ISO8601 format) of the source database</param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder SetAsPointInTimeRestore(this SqlDatabaseBuilder builder, Input<string> databaseId, Input<string> restorePointInTime)
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:165]: Should we rename this to "CreateAsPointInTimeRestore"? I feel the current name is a bit confusing.

File: src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs [hunk 4]
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
...
+        }
+
+        /// <summary>
+        /// Creates a database by restoring from a long term retention vault. 
+        /// recoveryServicesRecoveryPointResourceId must be specified as the recovery point resource ID.
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="databaseId"></param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder SetAsLongTermRetentionRestore(this SqlDatabaseBuilder builder, Input<string> databaseId)
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:180]: Should we rename this to "CreateAsLongTermRetentionRestore"? I feel the current name is a bit confusing.

File: src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs [hunk 5]
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
...
-        /// Restores the database from an existing one
+        /// Restores a deleted database using the database's original resource ID.
         /// </summary>
         /// <param name="builder"></param>
-        /// <param name="databaseId"></param>
+        /// <param name="databaseId">Original Resource ID of the database to restore from.</param>
+        /// <param name="deletionDate">Deletion date of the database in (ISO 8601 format).</param>
+        /// <param name="pointInTime">Point in time that you want to restore from (ISO 8601 format). Only specify to restore from an earlier point in time.</param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder SetAsRestore(this SqlDatabaseBuilder builder, Input<string> databaseId, Input<string> deletionDate,
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:99]: Should we rename this to "CreateAsRestoreOf"? I feel the current name is a bit confusing.

File: src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs [hunk 6]
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
...
-        //     Service Tiers](https://azure.microsoft.com/en-gb/documentation/articles/sql-database-service-tiers/).
+        /// Restores a deleted database using the Restorable Dropped Database Resource ID.
         /// </summary>
         /// <param name="builder"></param>
-        /// <param name="edition"></param>
+        /// <param name="databaseId">Resource ID of the Restorable Dropped Database to restore from</param>
+        /// <param name="pointInTime">Point in time that you want to restore from. Only specify to restore from an earlier point in time.</param>
         /// <returns></returns>
-        public static SqlDatabaseBuilder SkuTier(this SqlDatabaseBuilder builder, Input<string> edition)
+        public static SqlDatabaseBuilder SetAsRestore(this SqlDatabaseBuilder builder, Input<string> databaseId, Input<string> pointInTime = null)
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:75]: Should we rename this to "CreateAsRestoreOf"? I feel the current name is a bit confusing.

File: src/Infrastructure/Stize.Infrastructure.Azure.Tests/Sql/Stacks/DatabaseBasicStack.cs [hunk 7]
```diff
@@ -23,14 +24,36 @@ public DatabaseBasicStack()
...
+                .SkuTier("Basic")
+                .SkuServiceObjectiveName("S0")
+                .StorageAccountType(StorageAccountType.GRS)
+                .MaxDatabaseSizeGB(250)
+                .MinCapacity(100)
+                .DatabaseCollation("SQL_Latin1_General_CP1_CI_AS")
+                .SampleData(SampleName.AdventureWorksLT)
+                .Build();
+
+            var secondary = new SqlDatabaseBuilder("db1")
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure.Tests/Sql/Stacks/DatabaseBasicStack.cs:42]: Shouldn't this second builder be named "secondaryDB"?

**Review by sesispla — CHANGES_REQUESTED** [review]
Hi @LukeHarjulin!

Please find my feedback. I think we mainly need to make cosmetic changes to the method names.

File: src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs [hunk 8]
```diff
@@ -158,5 +278,80 @@ public static SqlDatabaseBuilder SkuFamily(this SqlDatabaseBuilder builder, Inpu
...
             return builder;
         }
+
+        /// <summary>
+        /// Provides enhanced availability by spreading replicas across availability zones within one region.
+        /// Sets the Zone Redundant property to 'true'.
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder EnableZoneRedundant(this SqlDatabaseBuilder builder)
```
Comment (RichardFowles89) [inline src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:288]: I think "EnableZoneRedundancy" might be a better name here

File: src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs (hunk 1 above)
Comment (RichardFowles89) [reply src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs]: Why did we change "RestoreFrom" as the name? RestoreFrom.. RecoverFrom are nicely fluent.

Comment (RichardFowles89) [reply src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs]: One other thought I had - should the database ID be passed in as the argument or the database object itself and then get the ID in the method? Passing in the db object might affect the performance a little bit but the method could read "RestoreFrom(database)"  -> just a thought

**Comment by LukeHarjulin** [comment]
I've updated method names and I have added support for the user to generate a new SQL server to host the new SQL DB (to solve the second part to the [issue](https://github.com/stize/infrastructure/issues/34) you raised @sesispla ).

**Review by sesispla — APPROVED** [review]

**Pull request merged**